apps/helix/parameters.txt
OX = oscillator

> On      - oscillator activity        - switch 2 (0-0.5 0.5-1)
> Amp     - oscillator amp             - bipolar from 0.5  (0-0.5 0.5=0db 0.5-1)
> Pan     - panning                    - 0..1
> HMCS    - ?                          - switch (?)
> Ocs     - octave                     - stepped -8 +8
> Semi    - semitones                  - stepped -24 +24
> Det     - cents                      - stepped -63 +63
> Phs     - phase                      - 0..1 (?)
> Free    - freerun                    - switch 2 (?)

> US      - multi oscillator width     - 0..1
> USS     - multi oscillator           - switch 2 (0-0.5 0.5-1)
> S-C     - super count                - switch 0..8
> S-P     - super phase                - 0..1
> S-F     - super free                 - 0..1
> S-W     - super wide                 - 0..1

> Form    - wavetype                   - 0 (wave)
                                         1 ()

> ShpTb   - shape 0 time bend          - 0..1
> ShpTm   - shape 0 time mirror        - 0..1
> ShpTw   - shape 0 time wrap          - 0..1
> ShpAm   - shape 0 amp mirror         - 0..1
> ShpAw   - shape 0 amp wrap           - 0..1
> ShpAb   - shape 0 amp bitreduce      - 0..1
> ShpAo   - shape 0 amp octave         - 0..1
> ShpAop  - shape 0 amp octave phase   - 0..1
> ShpAh   - shape 0 amp high pass      - 0..1
> ShpAl   - shape 0 amp low pass       - 0..1
> ShpAc   - shape 0 amp clip           - 0..1
> ShpFfb  - shape 0 freq fund boost    - 0..1

> ShpTb1  - shape 1 time bend          - 0..1
> ShpTm1  - shape 1 time mirror        - 0..1
> ShpTw1  - shape 1 time wrap          - 0..1
> ShpAm1  - shape 1 amp mirror         - 0..1
> ShpAw1  - shape 1 amp wrap           - 0..1
> ShpAb1  - shape 1 amp bitreduce      - 0..1
> ShpAo1  - shape 1 amp octave         - 0..1
> ShpAop1 - shape 1 amp octave phase   - 0..1
> ShpAh1  - shape 1 amp high pass      - 0..1
> ShpAl1  - shape 1 amp low pass       - 0..1
> ShpAc1  - shape 1 amp clip           - 0..1
> ShpFfb1 - shape 1 freq fund boost    - 0..1
> ShpFcp1 - shape 1 freq copy phases   - 0..1

> ShpMrp  - shape morph                - 0..1




global

> VA   - VCA




apps/capsaicin/parameters.txt
@fixes
- joystick behaves strange (flipping on automation)

@todo
- introduce compressor on part or master (i think is needed)
- disable all controls when the part is not enabled
- in the mixer, use some knobs to control whatever parameter the user wants
- add clipping indicator to master
- keep out part stereo selector
- keep out keyboard note range selector slider
- add part receive channel selector


========================================= effect types
master->sysefx[i]->seteffectpar (param, value)

0 - No efx -------------------------------------------
1 - reverb -------------------------------------------
    0 - volume
    1 - pan
    2 - time
    3 - l delay
    4 - l delay feedback
    5 - reflection delay
    6 - early refelction
    7 - lowpass
    8 - higpass
    9 - damp
    10 - type (0-random, 1-freeverb)
    11 - room size

2 - echo ---------------------------------------------
    0 - volume
    1 - pan
    2 - delay time
    3 - lr delay
    4 - lr c
    5 - feedback
    6 - damp

3 - chorus -------------------------------------------
    0 - volume
    1 - pan
    2 - frequency
    3 - random
    4 - lfo types (0-sine, 1-tri)
    5 - stereo image
    6 - depth
    7 - delay
    8 - feedback
    9 - l/r
    10 - ???
    11 - subtract (0..1)

4 - phaser -------------------------------------------
    0 - volume
    1 - pan
    2 - frequency
    3 - random
    4 - lfo types (0-sine, 1-tri)
    5 - stereo image
    6 - depth
    7 - feedback
    8 - stages (1..MAX_PHASER_STAGES)
    9 - l/r
    10 - subtract (0..1)
    11 - phase

5 - wah ----------------------------------------------
    0 - volume
    1 - pan
    2 - frequency
    3 - random
    4 - lfo types (0-sine, 1-tri)
    5 - stereo image
    6 - depth
    7 - feedback
    8 - delay (1..MAX_ALIENWAH_DELAY)
    9 - l/r
    10 - phase

6 - disto --------------------------------------------
    0 - volume
    1 - pan
    2 - lrc (?)
    3 - drive
    4 - level
    5 - type:
        0 - atan
        1 - asym1
        2 - pow
        3 - sine
        4 - quantize
        5 - zigzag
        6 - limit
        7 - limit U
        8 - limit L
        9 - ilmt
        10 - clip
        11 - asym2
        12 - pow2
        13 - segmentize
    6 - negative (0..1)
    7 - lowpass
    8 - highpass
    9 - stereo (0..1)
    10 - filter pre distortion (0..1)

7 - eq -----------------------------------------------
    0 - volume
    ---------- bands (0, MAX_EQ_BANDS - 1)
    (band * 5 + 10) - type:
                      0 - off
                      1 - lp1
                      2 - hp1
                      3 - lp2
                      4 - hp2
                      5 - bp2
                      6 - n2
                      7 - pk
                      8 - lsh
                      9 - hsh
    (band * 5 + 11) - frequency
    (band * 5 + 12) - gain
    (band * 5 + 13) - Q
    (band * 5 + 14) - stages (0, MAX_FILTER_STAGES - 1)

8 - filter -------------------------------------------
    0 - volume
    1 - pan
    2 - frequency
    3 - random
    4 - lfo types (0-sine, 1-tri)
    5 - stereo image
    6 - lfo depth
    7 - AS (?)
    8 - Amplitude modulation invert (0..1)
    9 - Amplitude modulation


======================================== master volume
    0..127
-------------------------------------------------------
    o->value(master->Pvolume);
-------------------------------------------------------
    master->setPvolume ((int) o->value());

======================================== fine detune
    0..127
-------------------------------------------------------
    o->value(master->microtonal.Pglobalfinedetune);
-------------------------------------------------------
    master->microtonal.Pglobalfinedetune = (int) o->value();

======================================== key shift
    -64..64
-------------------------------------------------------
    o->value (master->Pkeyshift - 64);
-------------------------------------------------------
    master->setPkeyshift ((int) o->value() + 64);

======================================== panic mode
-------------------------------------------------------
    pthread_mutex_lock (&master->mutex);
    master->shutup = 1;
    pthread_mutex_unlock (&master->mutex);

======================================== system effect
    o->value(master->sysefx[nsyseff]->geteffect());
-------------------------------------------------------
    pthread_mutex_lock(&master->mutex);
    master->sysefx[nsyseff]->changeeffect((int) o->value());
    pthread_mutex_unlock(&master->mutex);

======================================== insert effect
    o->value(master->insefx[ninseff]->geteffect());
    if (master->Pinsparts[ninseff]== -1) o->deactivate();
------------------------------------------------------
    pthread_mutex_lock(&master->mutex);
    master->insefx[ninseff]->changeeffect((int) o->value());
    pthread_mutex_unlock(&master->mutex);

======================================== insert effect to
    o->add("Master Out");o->add("Off");
    char tmp[50]; for (int i=0;i<NUM_MIDI_PARTS;i++) {sprintf(tmp,"Part %2d",i+1);o->add(tmp);};
    o->value(master->Pinsparts[ninseff]+2);
-------------------------------------------------------
    master->Pinsparts[ninseff]=(int) o->value()-2;
    if ((int) o->value()==1){
        inseffectuigroup->deactivate();
        insefftype->deactivate();
        inseffectui->deactivate();
    } else {
        inseffectuigroup->activate();
        insefftype->activate();
        inseffectui->activate();
    };
    master->insefx[ninseff]->cleanup();


======================================== part name
    (char *)part->Pname
    (int)   part->info.Ptype
    (char *)part->info.Pauthor
    (char *)part->info.Pcomments

======================================== part on off
    0..1
-------------------------------------------------------
    o->value(master->part[npart]->Penabled);
-------------------------------------------------------
    pthread_mutex_lock(&master->mutex);
    master->partonoff(npart,(int) o->value());
    pthread_mutex_unlock(&master->mutex);

======================================== part note on active
    0..1
-------------------------------------------------------
    o->value(part->Pnoteon);
-------------------------------------------------------
    part->Pnoteon=(int) o->value();

======================================== part poly
    0..1
-------------------------------------------------------
    o->value(part->Ppolymode);
-------------------------------------------------------
    part->Ppolymode=(int) o->value();

======================================== part poly
    0..1
-------------------------------------------------------
    o->value(part->Ppolymode);
-------------------------------------------------------
    part->Ppolymode=(int) o->value();

======================================== part volume
    0..127
-------------------------------------------------------
    o->value(master->part[npart]->Pvolume);
-------------------------------------------------------
    master->part[npart]->setPvolume((int) o->value());

======================================== part pan
    0..127
-------------------------------------------------------
    o->value(master->part[npart]->Ppanning);
-------------------------------------------------------
    master->part[npart]->setPpanning((int) o->value());

======================================== part velocity
    0..127
-------------------------------------------------------
    o->value(part->Pvelsns);
//    o->value(part->Pveloffs); // 0
-------------------------------------------------------
    part->Pvelsns=(int) o->value();
//    part->Pveloffs=(int) o->value(); // 0

======================================== part channel
    0..15
-------------------------------------------------------
    o->value(master->part[npart]->Prcvchn);
-------------------------------------------------------
    master->part[npart]->Prcvchn=(int) o->value();
    virkeys->midich=(int) o->value();

======================================== part min key
    0..127
-------------------------------------------------------
    o->value(master->part[npart]->Pminkey);
-------------------------------------------------------
    master->part[npart]->Pminkey=(int) o->value();

======================================== part max key
    0..127
-------------------------------------------------------
    o->value(master->part[npart]->Pmaxkey);
-------------------------------------------------------
    master->part[npart]->Pmaxkey=(int) o->value();

======================================== part shift
    -64..64
-------------------------------------------------------
    o->value(master->part[npart]->Pkeyshift-64);
-------------------------------------------------------
    master->part[npart]->Pkeyshift=(int) o->value()+64;

======================================== part portamento
    0..1
-------------------------------------------------------
    o->value(master->part[npart]->ctl.portamento.portamento);
    master->part[npart]->ctl.portamento.portamento=(int) o->value();

======================================== part bend range
    -6400..6400            step 100);
-------------------------------------------------------
    part->ctl.pitchwheel.bendrange=(int) o->value();

======================================== part midi controllers
    part->ctl.volume.receive = 0..1;
-------------------------------------------------------
    part->ctl.sustain.receive = 0..1;
    if (part->ctl.sustain.receive == 0) {
        part->RelaseSustainedKeys();
        part->ctl.setsustain(0);
    };
-------------------------------------------------------
    part->ctl.portamento.receive = 0..1;
    part->ctl.portamento.time = 0..127;
    part->ctl.portamento.updowntimestretch = 0..127;
    part->ctl.portamento.pitchthreshtype = 0..1;
    part->ctl.portamento.pitchthresh = 0..127;
-------------------------------------------------------
    part->ctl.fmamp.receive = 0..1
-------------------------------------------------------
    part->ctl.expression.receive = 0..1
-------------------------------------------------------
    part->ctl.modwheel.exponential = 0..1
    part->ctl.modwheel.depth = 0..127
-------------------------------------------------------
    part->ctl.bandwidth.exponential = 0..1
    part->ctl.bandwidth.depth = 0..127
-------------------------------------------------------
    part->ctl.filterq.depth = 0..127;
    part->ctl.filtercutoff.depth = 0..127;
    part->ctl.resonancebandwidth.depth = 0..127;
    part->ctl.resonancecenter.depth = 0..127;
    part->ctl.panning.depth = 0..127;

======================================== part effects
-------------------------------------------------------
    part->partefx[ninseff];
-------------------------------------------------------

======================================== part effect bypass
    0..1
-------------------------------------------------------
    int x=part->Pefxbypass[ninseff];o->value(x);
-------------------------------------------------------
    part->Pefxbypass[ninseff]=(((int)o->value())!=0);

======================================== part kit mode
    0..1
-------------------------------------------------------
    part->Pkitmode=(int) o->value();
    part->Pdrummode=(int) o->value();
    if (part->Pkitmode==0) {
        kitlist->deactivate();
    } else {
    kitlist->activate();
    };

======================================== part enabled
    part->kit[0].Padenabled = 0..1;
    part->kit[0].Psubenabled = 0..1;
    part->kit[0].Ppadenabled = 0..1;


apps/peggy2000/src/parameters.h
/*
 ==============================================================================

 This file is part of the JUCETICE project - Copyright 2008 by Lucio Asnaghi.

 JUCETICE is based around the JUCE library - "Jules' Utility Class Extensions"
 Copyright 2008 by Julian Storer.

 ------------------------------------------------------------------------------

 JUCE and JUCETICE can be redistributed and/or modified under the terms of
 the GNU Lesser General Public License, as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 JUCE and JUCETICE are distributed in the hope that they will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU Lesser General Public License
 along with JUCE and JUCETICE; if not, visit www.gnu.org/licenses or write to
 Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 Boston, MA 02111-1307 USA

 ==============================================================================

   @author  rockhardbuns
   @tweaker Lucio Asnaghi

 ==============================================================================
*/

#ifndef __JUCETICE_PEGGY2000_PARAMETERS_HEADER__
#define __JUCETICE_PEGGY2000_PARAMETERS_HEADER__

#include "StandardHeader.h"


class ParMan
{
public:

     enum
     {
         penv_a,
         penv_d,
         penv_s,
         penv_r,
         fenv_a,
         fenv_d, 
         fenv_s,
         fenv_r,
         aenv_a,
         aenv_d,
         aenv_s,
         aenv_r,
         aenv_amt,
         plfo_spd,
         plfo_amt,
         flfo_spd,
         flfo_amt,
         alfo_spd,
         alfo_amt,
         f_sat,
         f_res,
         f_cut,
         ch_spd,
         ch_dph,
         ch_mix,
         dl_tme,
         dl_lvl,
         dl_fb,
         g_sus,
         m_vol,
         g_swn,
         penv_amt,
         fenv_amt,
         g_len,
         f_80,
         vel_filter,
         vel_pitch,
         vel_amp,
         o1_cent,
         o2_cent,
         o_mix,
         o1_semi,
         o2_semi,
         o2_wave,
         o1_wave,
         g_tempo,
         g_mode,
         f_wave,
         plfo_wave,
         flfo_wave,
         alfo_wave,
         o2_env,
         o1_env,
         o1_lfo,
         o2_lfo,
         o1_sync,
         o_env,
         o_lfo,
         g_on,
         g_g1,
         g_g2,
         g_g3,
         g_g4,
         g_f1,
         g_f2,
         g_f3,
         g_f4,
         g_f5,
         g_f6,
         g_f7,
         g_f8,
         g_f9,
         g_f10,
         g_f11,
         g_f12,
         g_f13,
         g_f14,
         g_f15,
         g_f16
     };

     ParMan(){
          numPars = 81;
          pars = new float[numPars];
          loadPreset(0);
     }

     ~ParMan(){
          delete [] pars;
     }

     inline float getParameter(int i){
          i = jlimit(0, numPars, i);
          return pars[i];
     }

     int getNumPars(){
          return numPars;
     }

     void setParameter(int i, float v){
          i = jlimit(0, numPars, i);
          pars[i] = v;
     }

     bool loadPreset(int i = 0){
          if (i == 0){
               pars[0] = 0.000000f;
               pars[1] = 0.192871f;
               pars[2] = 0.000000f;
               pars[3] = 0.000000f;
               pars[4] = 0.000000f;
               pars[5] = 0.041608f;
               pars[6] = 0.000000f;
               pars[7] = 0.000000f;
               pars[8] = 0.000000f;
               pars[9] = 0.000000f;
               pars[10] = 1.000000f;
               pars[11] = 0.000000f;
               pars[12] = 0.340000f;
               pars[13] = 0.052000f;
               pars[14] = 0.011537f;
               pars[15] = 0.000000f;
               pars[16] = 0.000000f;
               pars[17] = 0.000000f;
               pars[18] = 0.000000f;
               pars[19] = 0.000000f;
               pars[20] = 0.120000f;
               pars[21] = 0.029057f;
               pars[22] = 0.164000f;
               pars[23] = 0.320000f;
               pars[24] = 0.516000f;
               pars[25] = 0.496000f;
               pars[26] = 0.296000f;
               pars[27] = 0.144000f;
               pars[28] = 0.532000f;
               pars[29] = 0.192000f;
               pars[30] = 0.000000f;
               pars[31] = 0.000000f;
               pars[32] = 0.168000f;
               pars[33] = 8.000000f;
               pars[34] = 0.624000f;
               pars[35] = 0.696000f;
               pars[36] = 0.000000f;
               pars[37] = 0.456000f;
               pars[38] = 0.000000f;
               pars[39] = 0.000000f;
               pars[40] = 0.000000f;
               pars[41] = 0.000000f;
               pars[42] = 0.000000f;
               pars[43] = 1.000000f;
               pars[44] = 1.000000f;
               pars[45] = 3.000000f;
               pars[46] = 1.000000f;
               pars[47] = 1.000000f;
               pars[48] = 1.000000f;
               pars[49] = 0.000000f;
               pars[50] = 0.000000f;
               pars[51] = 0.000000f;
               pars[52] = 0.000000f;
               pars[53] = 0.000000f;
               pars[54] = 1.000000f;
               pars[55] = 0.000000f;
               pars[56] = 0.000000f;
               pars[57] = 0.000000f;
               pars[58] = 0.000000f;
               pars[59] = 0.000000f;
               pars[60] = 0.000000f;
               pars[61] = 0.000000f;
               pars[62] = 0.000000f;
               pars[63] = 0.000000f;
               pars[64] = 0.000000f;
               pars[65] = 0.000000f;
               pars[66] = 0.000000f;
               pars[67] = 0.000000f;
               pars[68] = 0.000000f;
               pars[69] = 0.000000f;
               pars[70] = 0.000000f;
               pars[71] = 0.000000f;
               pars[72] = 0.000000f;
               pars[73] = 0.000000f;
               pars[74] = 0.000000f;
               pars[75] = 0.000000f;
               pars[76] = 0.000000f;
               pars[77] = 0.000000f;
               pars[78] = 0.000000f;
               pars[79] = 0.000000f;
               pars[80] = 0.000000f;
               return true;
          }

        return false;
     }

private:

     int numPars;
     float * pars;
};


#endif
